File: programmers/java/src/level2/kakao/kakao2018/NewsClustering.java
package level2.kakao.kakao2018;

import java.util.*;

public class NewsClustering {
    public static void main(String[] args) {
        NewsClustering nc = new NewsClustering();
        System.out.println(nc.solution("FRANCE", "french"));
        System.out.println(nc.solution_other("FRANCE", "french"));
    }
    public int solution(String str1, String str2) {
        List<String> str1List = new ArrayList<>();
        for (int i = 0; i < str1.length() -1; i++) {
            char c1 = str1.charAt(i);
            char c2 = str1.charAt(i+1);
            if(Character.isLetter(c1) && Character.isLetter(c2)) {
                str1List.add(str1.substring(i, i+2).toLowerCase());
            }
        }

        List<String> str2List = new ArrayList<>();
        for (int i = 0; i < str2.length() -1; i++) {
            char c1 = str2.charAt(i);
            char c2 = str2.charAt(i+1);
            if(Character.isLetter(c1) && Character.isLetter(c2)) {
                str2List.add(str2.substring(i, i+2).toLowerCase());
            }
        }

        int total = str1List.size() + str2List.size();
        int bunza = 0;
        if(total == 0) {
            return 65536;
        }
        int st2Idx =0;
        Collections.sort(str1List);
        Collections.sort(str2List);
        for (String s : str1List) {
            for (int i = st2Idx; i < str2List.size(); i++) {
                 if(str2List.get(i).equals(s)) {
                     st2Idx = i + 1;
                     bunza += 1;
                     total -= 1;
                     break;
                 }
            }
            if(st2Idx == str2List.size()) {
                break;
            }
        }
        return (int) (((float) bunza)/total * 65536);
    }
    public int solution_other(String str1, String str2) {
        List<String> str1List = new ArrayList<>();
        for (int i = 0; i < str1.length()-1; i++) {
            char c1 = str1.charAt(i);
            char c2 = str1.charAt(i+1);
            if(Character.isLetter(c1) && Character.isLetter(c2)) {
                str1List.add(str1.substring(i, i+2).toLowerCase());
            }
        }

        List<String> str2List = new ArrayList<>();
        for (int i = 0; i < str2.length()-1; i++) {
            char c1 = str2.charAt(i);
            char c2 = str2.charAt(i+1);
            if(Character.isLetter(c1) && Character.isLetter(c2)) {
                str2List.add(str2.substring(i, i+2).toLowerCase());
            }
        }
        Set<String> gyo = new HashSet<>();
        Set<String> hap = new HashSet<>();

        hap.addAll(str1List);
        hap.addAll(str2List);

        gyo.addAll(str1List);
        gyo.retainAll(str2List);

        if(hap.size() == 0) {
            return 65536;
        }

        int bunza = 0;
        int bunmo = 0;
        for(String g : gyo) {
            bunza += Math.min(Collections.frequency(str1List, g), Collections.frequency(str2List, g));
        }
        for(String h : hap) {
            bunmo += Math.max(Collections.frequency(str1List, h), Collections.frequency(str2List, h));
        }

        float float_bunza = (float) bunza;
        return (int) ((float_bunza / bunmo) * 65536);
    }
}

File: programmers/python/level2/kakao/2018/news_clustering.py
from collections import Counter

def solution(str1, str2):
    str1_list = []
    str2_list = []
    for i in range(len(str1)-1):
        s = str1[i:i+2]
        if s.isalpha():
            str1_list.append(s.lower())

    for i in range(len(str2)-1):
        s = str2[i:i+2]
        if s.isalpha():
            str2_list.append(s.lower())

    total = len(str1_list) + len(str2_list)
    bunza = 0
    if total == 0 :
        return 65536

    str1_list.sort()
    str2_list.sort()
    s2idx = 0 
    for i in range(len(str1_list)):
        s = str1_list[i]
        for j in range(s2idx, len(str2_list)):
            if s == str2_list[j]:
                s2idx = j+1
                bunza += 1
                total -= 1
                break
        if s2idx == len(str2_list):
            break
    return int(bunza/total * 65536)

def solution_old(str1, str2):
    st1 = []
    for i in range(len(str1)):
        if i == len(str1) - 1 :
            break
        c1 = str1[i]
        c2 = str1[i+1]
        if c1.isalpha() and c2.isalpha() :
            if c1.isupper():
                c1 = c1.lower()
            if c2.isupper():
                c2 = c2.lower()
            st1.append(c1+c2)
    
    st2 = []
    for i in range(len(str2)):
        if i == len(str2) - 1 :
            break
        c1 = str2[i]
        c2 = str2[i+1]
        if c1.isalpha() and c2.isalpha() :
            if c1.isupper():
                c1 = c1.lower()
            if c2.isupper():
                c2 = c2.lower()
            st2.append(c1+c2)

    st1_counter = Counter(st1)
    st2_counter = Counter(st2)

    bunza = 0
    bunmo = 0 
    for c1 in st1_counter:
        if c1 in st2_counter: 
            bunza += min(st1_counter[c1], st2_counter[c1])
            bunmo += max(st1_counter[c1], st2_counter[c1])
        else:
            bunmo += st1_counter[c1]
    
    for c2 in st2_counter:
        if not c2 in st1_counter:
            bunmo += st2_counter[c2]

    if bunmo == 0:
        return 65536
    return int(bunza / bunmo * 65536)

print(solution("FRANCE", "french"))
print(solution("handshake", "shake hands"))
print(solution("aa1+aa2", "AAAA12"))
print(solution("E=M*C^2", "e=m*c^2"))

print("aA".isalpha()) ##True

def solution2(str1, str2):
    str1 = [str1[n:n+2].lower() for n in range(len(str1)-1) if str1[n:n+2].isalpha()]
    str2 = [str2[n:n+2].lower() for n in range(len(str2)-1) if str2[n:n+2].isalpha()]

    gyo = set(str1) & set(str2)
    hap = set(str1) | set(str2)

    if len(hap) == 0 :
        return 65536

    gyo_sum = sum([min(str1.count(gg), str2.count(gg)) for gg in gyo])
    hap_sum = sum([max(str1.count(hh), str2.count(hh)) for hh in hap])

    return int((gyo_sum/hap_sum)*65536)
